streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala

```diff
@@ -82,13 +113,23 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
         "ReceiverTracker", new ReceiverTrackerEndpoint(ssc.env.rpcEnv))
       if (!skipReceiverLaunch) receiverExecutor.start()
       logInfo("ReceiverTracker started")
+      trackerState = Started
     }
   }
 
   /** Stop the receiver execution thread. */
   def stop(graceful: Boolean): Unit = synchronized {
-    if (!receiverInputStreams.isEmpty && endpoint != null) {
+    if (isTrackerStarted) {
       // First, stop the receivers
+      // acquire "stoppingTrackerLock" so that setting trackerState to "Stopping" and registering
+      // receivers won't happen at the same time
+      stoppingTrackerLock.synchronized {
+        trackerState = Stopping
+        if (!skipReceiverLaunch) {
+          // Send the stop signal to all the receivers
+          receiverExecutor.stopReceivers()
```

Comment:
It's weird. I think the abstraction of `ReceiverLauncher` has a little problem. `ReceiverLauncher. stopReceiver` actually will modify the state of `ReceiverTracker.receiverInfo`. So it forces me to expose `stopReceiver` to `ReceiverTracker`.